src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -69,6 +69,14 @@
   public static final Pattern GOOG_MODULE_EXTRACT =
       Pattern.compile("goog.module\\(['\"](.*)['\"]\\);");
 
+  /**
+   * Closure types can be used but not provided. Clutz collects them and
+   * visits them in a separate pass. A new set of types can be discovered
+   * while visiting unprovided types, thus recusively iterating. This holds
+   * the maximum depth the tool will visit.
+   */
+  private int maxTypeUsedDepth = 5;
```

Comment:
* this variable is mutable state on what is essentially the global object, and manipulated at a far distance. That's a recipe for nasty bugs in the future, in my experience.
* `5` is obviously an arbitrary number, so Clutz might still fail for some code.
* there's an odd collaboration going on between `processUnprovidedTypes` and `emitTemplatizedTypes` below.

I wonder how prevalent this code pattern is, it seems quite questionable. Can't we fix the code that exhibits the issue?

Alternatively: could we just emit all symbols, unconditionally? That seems to be conceptually what JSCompiler does behind the scenes. Why did we filter those again?